### ISSUE TITLE
cmsRun crash with multiple streams, a subprocess, and the summary

### ISSUE THREAD

**Issue by fwyzard** [cms-sw/cmssw#17045]
running a simple configuration file with
  - at least one non-empty subprocess
  - multiple streams (but not necessarily multiple threads)
  - process.options.wantSummary = True
makes cmsRun crash.

A sample configuration file is attached: [bug17045.py](https://github.com/cms-sw/cmssw/files/654272/bug17045.txt)

The job runs fine until the end, but then crashes with a segmentation fault while calling some destructors:
```
Starting program: .../bin/slc6_amd64_gcc620/cmsRun empty.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
%MSG-i ThreadSetup:  (NoModuleName) 13-Dec-2016 17:44:42 CET pre-events
setting # threads 1
%MSG
%MSG-i StreamSetup:  (NoModuleName) 13-Dec-2016 17:44:42 CET pre-events
setting # streams 4
%MSG
[New Thread 0x7fffe5d3f700 (LWP 132671)]
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 at 13-Dec-2016 17:44:49.865 CET
[New Thread 0x7fffe0e3f700 (LWP 132677)]
Begin processing the 2nd record. Run 1, Event 2, LumiSection 1 at 13-Dec-2016 17:44:49.866 CET
Begin processing the 3rd record. Run 1, Event 3, LumiSection 1 at 13-Dec-2016 17:44:49.866 CET
...
Begin processing the 100th record. Run 1, Event 100, LumiSection 1 at 13-Dec-2016 17:44:49.874 CET

TrigReport ---------- Event  Summary ------------
TrigReport Events total = 100 passed = 0 failed = 100
...
=============================================

MessageLogger Summary

Severity    # Occurrences   Total Occurrences
--------    -------------   -----------------

Thread 1 "cmsRun" received signal SIGSEGV, Segmentation fault.
0x00007ffff7a5fe97 in boost::detail::atomic_exchange_and_add (dv=-1, pw=0x11ad2837)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/boost/1.57.0/include/boost/smart_ptr/detail/sp_counted_base_gcc_x86.hpp:50
50          );
(gdb) bt
#0  0x00007ffff7a5fe97 in boost::detail::atomic_exchange_and_add (dv=-1, pw=0x11ad2837)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/boost/1.57.0/include/boost/smart_ptr/detail/sp_counted_base_gcc_x86.hpp:50
#1  boost::detail::sp_counted_base::release (this=0x11ad282f)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/boost/1.57.0/include/boost/smart_ptr/detail/sp_counted_base_gcc_x86.hpp:144
#2  boost::detail::shared_count::~shared_count (this=0x7fffe0eb2e50, __in_chrg=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/boost/1.57.0/include/boost/smart_ptr/detail/shared_count.hpp:443
#3  boost::shared_ptr<boost::re_detail::basic_regex_implementation<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > >::~shared_ptr (this=0x7fffe0eb2e48,
    __in_chrg=<optimized out>) at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/boost/1.57.0/include/boost/smart_ptr/shared_ptr.hpp:323
#4  boost::basic_regex<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::~basic_regex (this=0x7fffe0eb2e48, __in_chrg=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/boost/1.57.0/include/boost/regex/v4/basic_regex.hpp:346
#5  edm::ProductSelectorRules::Rule::~Rule (this=0x7fffe0eb2e40, __in_chrg=<optimized out>)
    at .../src/FWCore/Framework/interface/ProductSelectorRules.h:50
#6  std::_Destroy<edm::ProductSelectorRules::Rule> (__pointer=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:93
#7  std::_Destroy_aux<false>::__destroy<edm::ProductSelectorRules::Rule*> (__last=<optimized out>, __first=0x7fffe0eb2e40)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:103
#8  std::_Destroy<edm::ProductSelectorRules::Rule*> (__last=<optimized out>, __first=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:126
#9  std::_Destroy<edm::ProductSelectorRules::Rule*, edm::ProductSelectorRules::Rule> (__last=0x7fffe0eb2e88, __first=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:151
#10 std::vector<edm::ProductSelectorRules::Rule, std::allocator<edm::ProductSelectorRules::Rule> >::~vector (this=0x7fffe128db00, __in_chrg=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_vector.h:426
#11 edm::ProductSelectorRules::~ProductSelectorRules (this=0x7fffe128db00, __in_chrg=<optimized out>)
    at .../src/FWCore/Framework/interface/ProductSelectorRules.h:24
#12 edm::SubProcess::~SubProcess (this=0x7fffe128d800, __in_chrg=<optimized out>)
    at .../src/FWCore/Framework/src/SubProcess.cc:192
#13 0x00007ffff7add527 in std::_Destroy<edm::SubProcess> (__pointer=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:93
#14 std::_Destroy_aux<false>::__destroy<edm::SubProcess*> (__last=<optimized out>, __first=0x7fffe128d800)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:103
#15 std::_Destroy<edm::SubProcess*> (__last=<optimized out>, __first=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:126
#16 std::_Destroy<edm::SubProcess*, edm::SubProcess> (__last=0x7fffe128dbb0, __first=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_construct.h:151
#17 std::vector<edm::SubProcess, std::allocator<edm::SubProcess> >::~vector (this=0x7ffff221dd70, __in_chrg=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/stl_vector.h:426
#18 edm::EventProcessor::~EventProcessor (this=0x7ffff221dc00, __in_chrg=<optimized out>)
    at .../src/FWCore/Framework/src/EventProcessor.cc:569
#19 0x00007ffff7add859 in edm::EventProcessor::~EventProcessor (this=0x7ffff221dc00, __in_chrg=<optimized out>)
    at .../src/FWCore/Framework/src/EventProcessor.cc:585
#20 0x000000000040afb1 in std::default_delete<edm::EventProcessor>::operator() (this=<optimized out>, __ptr=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/unique_ptr.h:76
#21 std::unique_ptr<edm::EventProcessor, std::default_delete<edm::EventProcessor> >::~unique_ptr (this=0x7fffffff35f0, __in_chrg=<optimized out>)
    at /cvmfs/cms-ib.cern.ch/2016-51/slc6_amd64_gcc620/external/gcc/6.2.0/include/c++/6.2.0/bits/unique_ptr.h:236
#22 (anonymous namespace)::EventProcessorWithSentry::~EventProcessorWithSentry (this=0x7fffffff35f0, __in_chrg=<optimized out>)
    at .../src/FWCore/Framework/bin/cmsRun.cpp:66
#23 0x000000000040a9aa in main (argc=<optimized out>, argv=<optimized out>) at .../src/FWCore/Framework/bin/cmsRun.cpp:137
```

I can reproduce the problem with
  - CMSSW_9_0_0_pre1
  - CMSSW_9_0_X_2016-12-04-1100
  - CMSSW_9_0_X_2016-12-13-1100
  - CMSSW_9_0_DEVEL_X_2016-12-13-1100

I do not see the problem with CMSSW_8_1_0

**Comment by fwyzard** [comment]
A sample configuration file is attached: [bug17045.py](https://github.com/cms-sw/cmssw/files/654272/bug17045.txt)

**Comment by Dr15Jones** [comment]
assign core

**Comment by Dr15Jones** [comment]
@fwyzard thanks for making the issue

**Comment by Dr15Jones** [comment]
fixed in #17060

**Comment by Dr15Jones** [comment]
+1
